jni/cmake/init-simd.cmake

```diff
@@ -0,0 +1,107 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright OpenSearch Contributors
+
+include(CheckCXXSourceCompiles)
+
+# Allow user overrides
+if(NOT DEFINED AVX2_ENABLED)
+    set(AVX2_ENABLED ON)
+endif()
+
+if(NOT DEFINED AVX512_ENABLED)
+    set(AVX512_ENABLED ON)
+endif()
+
+if(NOT DEFINED AVX512_SPR_ENABLED)
+    set(AVX512_SPR_ENABLED OFF)
+endif()
+
+# SIMD feature flags default to OFF
+set(KNN_HAVE_F16C OFF)
+set(KNN_HAVE_AVX512 OFF)
+set(KNN_HAVE_ARM_FP16 OFF)
+
+set(SIMD_OPT_LEVEL "")
+set(SIMD_FLAGS "")
+
+# AVX512 detection (compiler only)
+if(AVX512_ENABLED)
+    set(CMAKE_REQUIRED_FLAGS "-mavx512f -mf16c")

```

Comment:
The whole detection logic is very interesting. But that is not how we would like to put this. So the way we should be thinking about these things is:
1. operator pass the flag which says enable SIMD_OPT for the knn build.
2. We either build the so file or we fail. This is similar to how faiss does. If some one pass avx_512 flag and system doesn't support it then build fails
3. When the whole k-NN plugin will be packaged we should build all the different version of opensearch_smid just like we do for faiss. 
4. At runtime we should check if SIMD with avx_512 is supported we pick the right so files.

Reply:
@navneet1v I added these changes in this PR itself since the fix for the MACOS test was overlapping with the above. Thanks!